doc/dvsource-dvgrab.1
.\" dvsource-dvgrab.1 written by Ben Hutchings <ben@decadent.org.uk>
.TH DVSOURCE-DVGRAB 1 "18 February 2009"
.SH NAME
dvsource-dvgrab \- Firewire and Video4Linux2 sources for DVswitch
.SH SYNOPSIS
.HP
.B dvsource-firewire
.RB [ \-h
.IR host ]
.RB [ \-p
.IR port ]
.RB [ \-c
.IR card | device ]
.HP
.B dvsource-v4l2-dv
.RB [ \-h
.IR host ]
.RB [ \-p
.IR port ]
.RI [ device ]
.SH DESCRIPTION
.LP
\fBdvsource-firewire\fR uses \fBdvgrab\fR(1) to stream from a DV
device (camera or VCR) connected by Firewire.
.LP
\fBdvsource-v4l2-dv\fR uses \fBdvgrab\fR(1) to stream from a DV device
that has a Video4Linux2 driver.  This includes most USB DV cameras if
you have the uvcvideo driver installed.
.SH OPTIONS
.TP
.BI \-h " host " \-p " port"
.RS
Specify the network address on which DVswitch is listening.  The host
address may be specified by name or as an IPv4 or IPv6 literal.
.RE
.TP
.BI \-c " card"
.RS
Specify the index of the Firewire card the DV device is connected to.
.RE
.SH AUTHOR
Ben Hutchings <ben@decadent.org.uk>.
.SH SEE ALSO
dvgrab(1), /usr/share/doc/dvswitch/README


ChangeLog
dvswitch (0.8) unstable; urgency=low

  * Added ticks and labels to the VU-meter
  * Bug fixes:
    - Removed use of function not present in Gtkmm 2.8
    - Fixed warnings from CMake 2.6
    - Fixed warnings from GCC
    - Fixed drag-selection when the pointer moves beyond the left or top
      edge of the display
    - Fixed memory corruption/crash in a selection corner-case
    - Fixed synchronisation of display buffers with the X server
    - Fixed repainting of main display before any frames have been received

 -- Ben Hutchings <ben@decadent.org.uk>  Wed, 04 Feb 2009 23:16:13 +0000

dvswitch (0.8~rc1) unstable; urgency=low

  * Disabled recording at startup
  * Added menu, currently with Quit command
  * Added buttons to GUI for effect selection
  * Changed key commands slightly
  * Allowed changing of primary and secondary video sources without
    cancelling the effect
  * Made dvsink-files create directories as necessary
  * Made dvsource-file reject non-DV files
  * Added support for USB DV sources through dvgrab and the uvcvideo driver
  * Moved source selector buttons to keep window within 1024x768 (depending
    on font selection and window decoration)
  * Added VU-meter for basic audio monitoring
  * Added warning indicator for sources which don't match the selected video
    and audio format
  * Documentation changes:
    - Documented all key commands
    - Changed contact address to dvswitch-devel@lists.alioth.debian.org
    - Added notes to (maybe) avoid user mistakes
  * Build system changes:
    - Changed to use cmake
    - Fixed building with current ffmpeg libavcodec library
  * Bug fixes:
    - Changed picture-in-picture selection to maintain frame aspect ratio
    - Restricted picture-in-picture to use the active area of the reduced
      frame
    - Slightly improved colour accuracy of picture-in-picture
    - Fixed cancellation of picture-in-picture selection
    - Fixed crash when user tries to apply picture-in-picture with an empty
      selection
    - Fixed crash when display depth is not 24 or 32
    - Fixed crash when users click in the dvswitch window without any source
      enabled
    - Made dvsource-file work with non-seekable files such as pipes

 -- Ben Hutchings <ben@decadent.org.uk>  Tue, 27 Jan 2009 01:04:43 +0000

dvswitch (0.7) unstable; urgency=low

  * Made keyboard shortcut for cut button work
  * Moved video/DV system parameters into a common structure
  * Replaced libdv codec with DV codec in libavcodec (ffmpeg), which
    is significantly faster
  * Added UI for picture-in-picture mixing

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 20 Apr 2008 22:18:37 +0100

dvswitch (0.6.2) unstable; urgency=low

  * Fixed bug in thumbnail image size initialisation that could result
    in a fatal BadValue error from X
  * Fixed minor bug in thumbnail image cleanup
  * Fixed bug that caused all sinks to be disabled when not recording

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 24 Feb 2008 10:55:41 +0100

dvswitch (0.6.1) unstable; urgency=low

  * Improved X driver compatibility:
    - Fixed calculation of Xv port numbers
    - Changed selection of visual class and depth for images to match the
      window
    - Added error messages and later checks for failure of image creation,
      where previously dvswitch could crash

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 21 Feb 2008 22:17:32 +0000

dvswitch (0.6) unstable; urgency=low

  * Added some infrastructure for video effects including
    picture-in-picture mixing
  * Switched from gtkmm 2.0/2.2 to 2.4+
  * Corrected small error in video display width
  * Changed the server to indicate overflow in output to sinks rather
    than disconnecting sinks that have overflowed; the file sink will
    create a new file after each overflow
  * Implemented dynamic display sizing, adding support for 16:9 frames
  * Turned source selection images into radio buttons
  * Fixed labelling of sources numbered above 9
  * Added support for starting and stopping recording while leaving
    any stream running
  * Added control buttons for recording and changed to default
    mnemonic for Cut

 -- Ben Hutchings <ben@decadent.org.uk>  Tue, 19 Feb 2008 09:37:22 +0000

dvswitch (0.5.1) unstable; urgency=low

  * Replaced FIREWIRE_DEVICE and --device option to dvsource-firewire
    with FIREWIRE_CARD and --card option
  * Corrected arguments to dvgrab

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 10 Jun 2007 21:41:17 +0100

dvswitch (0.5.0) unstable; urgency=low

  * Added indicator of title-safe area

 -- Ben Hutchings <ben@decadent.org.uk>  Mon, 30 Apr 2007 02:23:08 +0100

dvswitch (0.4) unstable; urgency=low

  * Implemented a command-spawning sink
  * Improved error handling
  * Added ability to switch between audio sources
  * Added indicators of selected video and audio sources

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 15 Apr 2007 16:27:47 +0100

dvswitch (0.3) unstable; urgency=low

  * Implemented dubbing of audio source onto video source (and
    silencing if audio source is unavailable)
  * Implemented reuse of source slots, so a disconnected source that
    reconnects will normally reappear in the same place
  * Added support for digit keys on the numeric keypad
  * Fixed priority of display updates, so the user interfaceshould
    remain responsive even with silly numbers of sources
  * Implemented synchronisation of mixer clock with audio source clock
  * Implemented sink serving and a DIF file sink
  * Fixed various bugs

 -- Ben Hutchings <ben@decadent.org.uk>  Tue, 10 Apr 2007 01:54:32 +0100

dvswitch (0.2) unstable; urgency=low

  * Fixed hang on exit
  * Implemented thumbnail displays of all sources
  * Switched to 1-based source selection in order to match keyboard layout
  * Reduced display lag
  * Reduced processor load of display updates by using Xv and Xshm

 -- Ben Hutchings <ben@decadent.org.uk>  Fri,  6 Apr 2007 00:54:22 +0100

dvswitch (0.1) unstable; urgency=low

  * Initial release


src/dvsource-dvgrab.c
/* Copyright 2007-2008 Ben Hutchings.
 * Copyright 2008 Petter Reinholdtsen.
 * See the file "COPYING" for licence details.
 */

#include <assert.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <getopt.h>
#include <unistd.h>

#include "config.h"
#include "protocol.h"
#include "socket.h"

static struct option options[] = {
    {"card",     1, NULL, 'c'},
    {"firewire", 0, NULL, 'F'},
    {"v4l2",     0, NULL, 'V'},
    {"host",     1, NULL, 'h'},
    {"port",     1, NULL, 'p'},
    {"help",     0, NULL, 'H'},
    {NULL,       0, NULL, 0}
};

enum mode {
    mode_unknown,
    mode_firewire,
    mode_v4l2,
};

static enum mode mode = mode_unknown;
static char * device_name = NULL;
static char * firewire_card = NULL;
static char * mixer_host = NULL;
static char * mixer_port = NULL;

static enum mode program_mode(const char * progname)
{
    const char * slash = strrchr(progname, '/');

    if (slash)
	progname = slash + 1;

    if (strcmp(progname, "dvsource-firewire") == 0)
	return mode_firewire;
    if (strcmp(progname, "dvsource-v4l2-dv") == 0)
	return mode_v4l2;
    return mode_unknown;
}

static void handle_config(const char * name, const char * value)
{
    if (strcmp(name, "FIREWIRE_CARD") == 0)
    {
	free(firewire_card);
	firewire_card = strdup(value);
    }
    else if (strcmp(name, "FIREWIRE_DEVICE") == 0 && mode == mode_firewire)
    {
	free(device_name);
	device_name = strdup(value);
    }
    else if (strcmp(name, "V4L2_DV_DEVICE") == 0 && mode == mode_v4l2)
    {
	free(device_name);
	device_name = strdup(value);
    }
    else if (strcmp(name, "MIXER_HOST") == 0)
    {
	free(mixer_host);
	mixer_host = strdup(value);
    }
    else if (strcmp(name, "MIXER_PORT") == 0)
    {
	free(mixer_port);
	mixer_port = strdup(value);
    }
}

static void usage(const char * progname)
{
    static const char firewire_args[] =
	"[{-c|--card} FIREWIRE-CARD] | DEVICE-NAME]";
    static const char v4l2_args[] = "[DEVICE-NAME]";
    static const char network_args[] =
	"[{-h|--host} MIXER-HOST] [{-p|--port} MIXER-PORT]";

    switch (program_mode(progname))
    {
    case mode_unknown:
	fprintf(stderr,
		"Usage: %s %s \\\n"
		"           --firewire %s\n"
		"       %s %s \\\n"
		"           --v4l2 %s\n",
		progname, network_args, firewire_args,
		progname, network_args, v4l2_args);
	break;
    case mode_firewire:
	fprintf(stderr,
		"Usage: %s %s \\\n"
		"           %s\n",
		progname, network_args, firewire_args);
	break;
    case mode_v4l2:
	fprintf(stderr,
		"Usage: %s %s \\\n"
		"           %s\n",
		progname, network_args, v4l2_args);
	break;
    }
}

int main(int argc, char ** argv)
{
    mode = program_mode(argv[0]);

    /* Initialise settings from configuration files. */
    dvswitch_read_config(handle_config);

    /* Parse arguments. */

    int opt;
    while ((opt = getopt_long(argc, argv, "c:h:p:", options, NULL)) != -1)
    {
	switch (opt)
	{
	case 'c':
	    free(firewire_card);
	    firewire_card = strdup(optarg);
	    break;
	case 'F':
	    mode = mode_firewire;
	    break;
	case 'V':
	    mode = mode_v4l2;
	    break;
	case 'h':
	    free(mixer_host);
	    mixer_host = strdup(optarg);
	    break;
	case 'p':
	    free(mixer_port);
	    mixer_port = strdup(optarg);
	    break;
	case 'H': /* --help */
	    usage(argv[0]);
	    return 0;
	default:
	    usage(argv[0]);
	    return 2;
	}
    }

    if (optind != argc)
    {
	free(device_name);
	device_name = strdup(argv[optind++]);
    }

    if (!mixer_host || !mixer_port)
    {
	fprintf(stderr, "%s: mixer hostname and port not defined\n",
		argv[0]);
	return 2;
    }

    if (optind != argc)
    {
	fprintf(stderr, "%s: excess argument \"%s\"\n",
		argv[0], argv[optind]);
	usage(argv[0]);
	return 2;
    }

    if (mode == mode_firewire)
    {
	if (device_name)
	    printf("INFO: Reading from Firewire device %s\n", device_name);
	else if (firewire_card)
	    printf("INFO: Reading from Firewire card %s\n", firewire_card);
	else
	    printf("INFO: Reading from first Firewire card with camera\n");
    }
    else if (mode == mode_v4l2)
    {
	if (!device_name)
	    device_name = "/dev/video";
	printf("INFO: Reading from V4L2 device %s\n", device_name);
    }
    else
    {
	fprintf(stderr, "%s: mode not defined (Firewire or V4L2)\n", argv[0]);
	return 2;
    }

    /* Connect to the mixer, set that as stdout, and run dvgrab. */

    printf("INFO: Connecting to %s:%s\n", mixer_host, mixer_port);
    int sock = create_connected_socket(mixer_host, mixer_port);
    assert(sock >= 0); /* create_connected_socket() should handle errors */
    if (write(sock, GREETING_SOURCE, GREETING_SIZE) != GREETING_SIZE)
    {
	perror("ERROR: write");
	exit(1);
    }
    if (dup2(sock, STDOUT_FILENO) < 0)
    {
	perror("ERROR: dup2");
	return 1;
    }
    close(sock);

    char * dvgrab_argv[7];
    char ** argp = dvgrab_argv;
    *argp++ = "dvgrab";
    if (mode == mode_v4l2)
	*argp++ = "-v4l2";
    if (device_name)
    {
	*argp++ = "-input";
	*argp++ = device_name;
    }
    else if (firewire_card)
    {
	*argp++ = "-card";
	*argp++ = firewire_card;
    }
    *argp++ = "-noavc";
    *argp++ = "-";
    *argp = NULL;
    assert(argp < dvgrab_argv + sizeof(dvgrab_argv) / sizeof(dvgrab_argv[0]));

    execvp("dvgrab", dvgrab_argv);
    perror("ERROR: execvp");
    return 1;
}
